Documentation on napari types in magicgui (#2306)
Napari/napari: docs/guides/magicgui.md

---
jupytext:
  cell_metadata_filter: -all
  formats: md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.12
    jupytext_version: 1.8.2
kernelspec:
  display_name: Python 3
  language: python
  name: python3
---

> (magicgui)=

Reviewer: What does this do?

Author: that's a "[target header](https://jupyterbook.org/reference/cheatsheet.html#target-headers)" (used for external linking purposes).  I saw that @kne42 had used them in other files, so I followed suit ☺️

> (magicgui)=

Reviewer 2: this line shows up when I preview the markdown file in VSCode - is that supposed to happen? Is it just a vscode quirk?

Author: yeah, the VSCode preview engine probably isn't running the whole myst-md parser that will be used when transpiling this document in jupyter-books, so a lot of it will look different.

python types to widgets.  In particular, it makes building widgets to represent
function inputs easy:

```{code-cell} python
from magicgui import magicgui
import datetime
import pathlib

@magicgui(
    call_button="Calculate",
    slider_float={"widget_type": "FloatSlider", 'max': 10},
    dropdown={"choices": ['first', 'second', 'third']},
)
def widget_demo(
    maybe: bool,
    some_int: int,
    spin_float=3.14159,
    slider_float=4.5,
    string="Text goes here",
    dropdown='first',
    date=datetime.datetime.now(),
    filename=pathlib.Path('/some/path.ext')
):
    ...

widget_demo.show()
```

For more information on the features and usage of `magicgui`, see the [magicgui
documentation](https://napari.org/magicgui).  `magicgui` does not require
napari, but napari *does* provide support for using magicgui within napari. The
purpose of this page is to document some of the conveniences provided by napari
when using `magicgui` with napari-specific type annotations.

## magicgui and type annotations

`magicgui` uses [type hints](https://www.python.org/dev/peps/pep-0484/) to infer
the appropriate widget type for a given function parameter, and to indicate a

> context-dependent action for the object returned from the function. (in the

Reviewer 2: ```suggestion
context-dependent action for the object returned from the function (in the
```

absense of a type hint, the type of the default value will be used).  Third
party packages (like `napari` in this case) may provide support for their types
using
[`magicgui.register_type`](https://napari.org/magicgui/usage/types_widgets.html#register-type).

> This is how using the type annotations described below lead to widgets and/or

Reviewer 2: ```suggestion
This is how using the type annotations described below leads to widgets and/or
```

"actions" in napari.

```{important}

> All of the type annotations described below *require* that the resulting widget
> be added to a napari viewer (using, e.g., `viewer.window.add_dock_widget`, or
> providing a magicgui-based widget via the {func}`~napari.plugins.hook_specifications.napari_experimental_provide_dock_widget` plugin hook specification.)

Reviewer 2: ```suggestion
All of the type annotations described below *require* that the resulting widget
be added to a napari viewer (either via `viewer.window.add_dock_widget`, or
by providing a magicgui-based widget via the {func}`~napari.plugins.hook_specifications.napari_experimental_provide_dock_widget` plugin hook specification).
```

```

## Parameter Annotations

The following napari types may be used as *parameter* type annotations in
magicgui functions to get information from the napari viewer into your
magicgui function. The consequence of each type annotation is described below:

- any napari {class}`~napari.layers.Layer` type, such as
  {class}`~napari.layers.Image` or {class}`~napari.layers.Points`
- any of the `<LayerType>Data` types from {mod}`napari.types`, such as
  {attr}`napari.types.ImageData` or  {attr}`napari.types.LabelsData`
- {class}`napari.Viewer`

> ### Annotating as a `Layer` subclasses

Reviewer 2: ```suggestion
### Annotating as a `Layer` subclass
```

viewer.window.add_dock_widget(my_widget)
```

*Note the widget at the bottom with "My Image" as the currently selected option*

```{code-cell} python
:tags: [remove-input]
from napari.utils import nbscreenshot

viewer.window._qt_window.resize(750, 550)
nbscreenshot(viewer)
```

### Annotating as `Layer`

In the previous example, the dropdown menu will *only* show
{class}`~napari.layers.Image` layers, because the parameter was annotated as an
{class}`~napari.layers.Image`.  If you'd like a dropdown menu that allows the
user to pick from *all* layers in the layer list, annotate your parameter as
{class}`~napari.layers.Layer`

```python
from napari.layers import Layer

@magicgui
def my_widget(layer: Layer):
    # do something with whatever layer the user has selected
    # note: it *may* be None! so your function should handle the null case
    ...
```

### Annotating as `napari.types.*Data`

In the previous example, the object passed to your function will be the actual
{class}`~napari.layers.Layer` instance, meaning you will need to access any
attributes (like `layer.data`) on your own.  If your function is designed to

> accept a numpy array, you can use the any of the special `<LayerType>Data` types

Reviewer 2: ```suggestion
accept a numpy array, you can use any of the special `<LayerType>Data` types
```

from {mod}`napari.types` to indicate that you only want the data attribute from
the layer (where `<LayerType>` is one of the available layer types).  Here's an
example using {attr}`napari.types.ImageData`

```python
from napari.types import ImageData
import numpy as np

@magicgui
def my_widget(array: ImageData):
    # note: it *may* be None! so your function should handle the null case
    if array is not None:
      assert isinstance(array, np.ndarray)  # it will be!
```

### Annotating as `napari.Viewer`

Lastly, if you need to access the actual {class}`~napari.viewer.Viewer` instance
in which the widget is docked, you can annotate one of your parameters as a
{class}`napari.Viewer`.

```python
from napari import Viewer

@magicgui
def my_widget(viewer: Viewer):
  ...
```

```{caution}

> Please use this sparingly, as a last resort. If you need to *add* layers

Reviewer 2: Can we give an example where it would be reasonable to provide the Viewer? Are there such examples?

Author: I hesitate to give an example, because of how frequently people see an example and just copy/paste it assuming it's best practice.  But, this would be necessary if you needed one of the viewer attributes that didn't exist on the layers themselves (like theme, ndisplay, grid, perhaps viewer.status, etc...).  Again, it's nothing we want to promote as a "go to" API, but it's there because people have needed it.  See for example, [usage in affinder](https://github.com/jni/affinder/blob/b538689a92ce878efae8316cb981fd0c9fb9ced8/affinder/affinder.py#L96), where it was needed to move layers around and adjust selection state.

Reviewer 3: Yeah, I was wondering about this bit too. We can probably leave for now, but we might want to update as we get more experience with when this is necessary etc.

Reviewer 3: or when we can suggest other alternatives etc.

to the viewer from your function, prefer one of the return-annotation methods
described [below](#adding-layers-to-napari-from-your-magicgui-function).
If you find that you require the viewer instance because of functionality that
is otherwise missing here, please consider opening an issue in the
[napari issue tracker](https://github.com/napari/napari/issues/new/choose),
describing your use case.
```

## Return Annotations

> The following napari types may be used as *return* type annotations in magicgui

Reviewer 2: ```suggestion
The following napari types may be used as *return* type annotations in `magicgui`
```

> functions to add layers to napari from your magicgui function. The consequence of

Reviewer 2: ```suggestion
functions to add layers to napari from your `magicgui` function. The consequence of
```

  return (data, {'properties': props}, 'points')

viewer = napari.Viewer()
viewer.window.add_dock_widget(make_points)
make_points()  # "call the widget" to call the function.
               # Normally this would be caused by some user UI interaction
```

```{code-cell} python
:tags: [remove-input]
from napari.utils import nbscreenshot

viewer.window._qt_window.resize(750, 550)
nbscreenshot(viewer)
```

### Returning `List[napari.types.LayerDataTuple]`

You can also create multiple layers by returning a list of
{attr}`~napari.types.LayerDataTuple`.

```python
from typing import List

@magicgui
def make_points(...) -> List[napari.types.LayerDataTuple]:
  ...
```

```{note}
Note: the `List[]` syntax here is optional from the perspective of `napari`.  You
can return either a single tuple or a list of tuples and they will all be added
to the viewer as long as you use either `List[napari.types.LayerDataTuple]` or 
`napari.types.LayerDataTuple`.  If you want your code to be properly typed, however,
your return type must match your return annotation.
```

### Updating an existing Layer

The default behavior is to add a new layer to the viewer for each

> `LayerDataTuple` returned by a magicgui function. By providing your a unique

Reviewer 2: ```suggestion
`LayerDataTuple` returned by a magicgui function. By providing a unique
```

`name` key in your {attr}`~napari.types.LayerDataTuple` metadata dict, you can
update an existing layer, rather than creating a new layer each time the
function is called:

```{code-cell} python
:tags: [remove-output]

@magicgui(call_button='Make Points', n_points={'maximum': 200})
def make_points(n_points=40) -> napari.types.LayerDataTuple:
  data = 500 * np.random.rand(n_points, 2)
  return (data, {'name': 'My Points'}, 'points')

viewer = napari.Viewer()
viewer.window.add_dock_widget(make_points)
# calling this multiple times will just update 'My Points'
make_points()
make_points.n_points.value = 80
make_points()
make_points.n_points.value = 120
make_points()
```

```{code-cell} python
:tags: [remove-input]
from napari.utils import nbscreenshot

viewer.window._qt_window.resize(750, 550)
nbscreenshot(viewer)
```

## Avoid imports with forward references

> Sometimes, it is undesireable to import and/or depend on `napari` directly just

Reviewer 2: ```suggestion
Sometimes, it is undesirable to import and/or depend on `napari` directly just
```

could be provided as a napari plugin as follows:

```python
from magicgui import magic_factory
from napari_plugin_engine import napari_hook_implementation

@magic_factory(auto_call=True, threshold={'max': 2 ** 16})
def threshold(
    data: 'napari.types.ImageData', threshold: int
) -> 'napari.types.LabelsData':
    return (data > threshold).astype(int)

@napari_hook_implementation
def napari_experimental_provide_dock_widget():
    return threshold
```

:::{note}
{func}`@magic_factory <magicgui.magic_factory>` behaves very much like
{func}`functools.partial`: it returns a callable that "remembers" some or
all of the parameters required for a "future" call to {func}`magicgui.magicgui`.
The parameters provided to {func}`@magic_factory <magicgui.magic_factory>` can
also be overridden when creating a widget from a factory:

```python
@magic_factory(call_button=True)
def my_factory(x: int):
    ...

widget1 = my_factory()
widget2 = my_factory(call_button=False, x={'widget_type': 'Slider'})
```

:::

### `magicgui.widgets.FunctionGui`

The other option for using `magicgui` in plugins is to directly subclass
{class}`magicgui.widgets.FunctionGui` (which is the type that is returned

> by the {func}`@magicgui <magicgui.magicgui>` decorator.)

Reviewer 2: ```suggestion
by the {func}`@magicgui <magicgui.magicgui>` decorator).
```
